src/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs

```diff
@@ -54,33 +54,29 @@ internal UnboundedChannelReader(UnboundedChannel<T> parent)
 
             public override Task Completion => _parent._completion.Task;
 
-            public override ValueTask<T> ReadAsync(CancellationToken cancellationToken) =>
-                TryRead(out T item) ?
-                    new ValueTask<T>(item) :
-                    ReadAsyncCore(cancellationToken);
-
-            private ValueTask<T> ReadAsyncCore(CancellationToken cancellationToken)
+            public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
             {
                 if (cancellationToken.IsCancellationRequested)
                 {
                     return new ValueTask<T>(Task.FromCanceled<T>(cancellationToken));
                 }
 
+                // Dequeue an item if we can.
                 UnboundedChannel<T> parent = _parent;
+                if (parent._items.TryDequeue(out T item))
```

Comment:
>if (parent._items.TryDequeue(out T item)) [](start = 16, length = 41)

is this safe to do outside the lock?

Reply:
Yes.  _items is a concurrent queue, so it's safe to read from without a lock (note we were already doing this inside of TryRead), and if we successfully get an item, we can just return.  However, if we don't, we have to test it again once holding the lock.